File: Lab 4/D/Bob.java
import java.io.*;
import java.net.*;

public class Bob {
    public static void main(String[] args) {
        try {
            Socket socket = new Socket("localhost", 12345);

            // Receive public key from Eve
            int yd1 = Integer.parseInt(new BufferedReader(new InputStreamReader(socket.getInputStream())).readLine());
            System.out.println("Received public key for Bob: " + yd1);

            int p = 29; // prime number
            int a = 2; // primitive root

            // Generating public key of Bob and sending it to Alice
            System.out.print("Enter private key of Bob: ");
            int xb = Integer.parseInt(new BufferedReader(new InputStreamReader(System.in)).readLine());
            int yb = (int) Math.pow(a, xb) % p; // Public Key of Bob
            System.out.println("Public key of Bob is: " + yb);
            new PrintStream(socket.getOutputStream()).println(yb);

            // Calculate the secret key
            int k1 = (int) Math.pow(yd1, xb) % p;
            System.out.println("Secret key decrypted by Bob: " + k1);

            socket.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Lab 4/Py/Bob.py
import socket
port = 12345
client_socket = socket.socket()
client_socket.connect(('localhost', port))

                                            #<3 Generating public key of Bob and sending it to Alice (which would be intruded)>
yd1 = int(client_socket.recv(1024).decode())
print("Received public key for Bob:", yd1)
xb = int(input("Enter private key of Bob:"))
yb = pow(2,xb,29) #Public Key of BOB
print(f"Public key of Bob is: {yb}")

                                            #4. Calculating the secret key
k1 = pow(yd1,xb,29)
print(f"Secret key decrypted by Bob: {k1}")
client_socket.send(str(yb).encode())  
client_socket.close()
